proc/coredump.c

```diff
@@ -0,0 +1,518 @@
+{
+	const u32 zero = 0;
+	Elf_Nhdr nhdr;
+	size_t i;
+	elf_prstatus prstatus;
+	hal_memset(&prstatus, 0, sizeof(prstatus));
+	for (i = 0; i < threadCnt; i++) {
+		nhdr.n_namesz = sizeof(PRSTATUS_NAME);
+		nhdr.n_descsz = sizeof(elf_prstatus) + SIZE_COREDUMP_GREGSET;
+		nhdr.n_type = NT_PRSTATUS;
+		coredump_encodeChunk(state, (u8 *)&nhdr, sizeof(nhdr));
+		coredump_encodeChunk(state, (u8 *)PRSTATUS_NAME, sizeof(PRSTATUS_NAME));
+		/* alignment */
+		coredump_encodeChunk(state, (u8 *)&zero, align4(sizeof(PRSTATUS_NAME)) - sizeof(PRSTATUS_NAME));
+
+		prstatus.pr_pid = threads[i].tid;
+		coredump_encodeChunk(state, (u8 *)&prstatus, offsetof(elf_prstatus, pr_reg));
+		hal_coredumpGRegset(buff, threads[i].userContext);
+		coredump_encodeChunk(state, (u8 *)buff, SIZE_COREDUMP_GREGSET);
+		coredump_encodeChunk(state, (u8 *)&prstatus.pr_reg, sizeof(prstatus) - offsetof(elf_prstatus, pr_reg));
+
+		hal_coredumpThreadAux(buff, threads[i].userContext);
+		coredump_encodeChunk(state, (u8 *)buff, SIZE_COREDUMP_THREADAUX);
+	}
+}
+
+
+static size_t coredump_stackSize(void *userSp, process_t *process)
+{
+	map_entry_t t;
+	map_entry_t *e;
+	size_t stackSize;
+
+	t.vaddr = userSp;
+	t.size = 1;
+
+	proc_lockSet(&process->mapp->lock);
+
+	e = lib_treeof(map_entry_t, linkage, lib_rbFind(&process->mapp->tree, &t.linkage));
+	if (e == NULL) {
+		proc_lockClear(&process->mapp->lock);
+		return 0;
+	}
+	stackSize = e->vaddr + e->size - userSp;
+
+	proc_lockClear(&process->mapp->lock);
+
+	return stackSize;
+}
+
+
+static void coredump_dumpStack(process_t *process, cpu_context_t *ctx, coredump_state_t *state)
+{
+	void *userSp;
+	size_t stackSize;
+
+	userSp = hal_cpuGetUserSP(ctx);
+	stackSize = coredump_stackSize(userSp, process);
+
+	coredump_encodeChunk(state, userSp, stackSize);
+}
+
+
+static void coredump_dumpPhdr(u32 type, size_t offset, void *vaddr, size_t size, unsigned short flags, coredump_state_t *state)
+{
+	Elf_Phdr phdr;
+	phdr.p_type = type;
+	phdr.p_offset = offset;
+	phdr.p_vaddr = (Elf_Addr)vaddr;
+	phdr.p_paddr = 0;
+	phdr.p_filesz = size;
+	phdr.p_memsz = size;
+
+	phdr.p_flags = 0;
+	if (flags & PROT_READ) {
+		phdr.p_flags |= PF_R;
+	}
+	if (flags & PROT_WRITE) {
+		phdr.p_flags |= PF_W;
+	}
+	if (flags & PROT_EXEC) {
+		phdr.p_flags |= PF_X;
+	}
+
+	phdr.p_align = 0;
+	coredump_encodeChunk(state, (u8 *)&phdr, sizeof(phdr));
+}
+
+
+static void coredump_dumpPhdrs(coredump_threadinfo_t *threadInfo, size_t threadCnt, size_t segCnt, process_t *process, cpu_context_t *ctx, coredump_state_t *state)
+{
+	const size_t THREAD_NOTES_SIZE = sizeof(Elf_Nhdr) +
+			align4(sizeof(PRSTATUS_NAME)) +
+			sizeof(elf_prstatus) +
+			SIZE_COREDUMP_GREGSET +
+			SIZE_COREDUMP_THREADAUX;
+	const size_t NOTES_SIZE = SIZE_COREDUMP_GENAUX + threadCnt * (THREAD_NOTES_SIZE);
+	size_t currentOffset;
+	size_t stackSize;
+	void *userSp;
+	map_entry_t *e;
+	size_t i;
```

Comment:
In some `#define` configurations, some variables and functions in this file are unused - please check compilation warnings with different cases

Reply:
Casting to `(void)` in `else` path is ok

Reply:
It gets a little dense with all these defines. 
Maybe it'll be better to create some static constants at the beginning of the file writing regular if statements?
Ultimately, we want to use some kind of runtime config.

Reply:
Yeah, sounds good to me